### PR TITLE
feat: OpenId4VC support

### PR DESCRIPTION
**Pull request by 2mau**

### CODE REVIEW

File: packages/openid4vc-client/src/utils/metadata.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    credentialSubject: IssuerCredentialSubject
+  }
+  issuer: {
+    display?: MetadataDisplay[]
+    id: string
+  }
+}
+
+// what does this mean
+const openId4VcCredentialMetadataKey = '_paradym/openId4VcCredentialMetadata'
```
Comment (berendsliedrecht) [inline packages/openid4vc-client/src/utils/metadata.ts:26]: This is a storage key that can be used to retrieve metadata about the credential. Likely for ease of access for displaying. @TimoGlastra likely knows whether we have to keep an equivalent or if this is up to the user to define.

Comment (2mau) [reply packages/openid4vc-client/src/utils/metadata.ts:26]: The PR is not yet finished. Just pushed some progress

Comment (TimoGlastra) [reply packages/openid4vc-client/src/utils/metadata.ts:26]: This is the display / issuer metadata. I chose to not add it as PR in AFJ as the spec is continuously evolving and thus it will contain different structures in the metadata. Not sure if that's a problem. It is very useful to have this though, but it may be better if we can generalize it and create a universal credential display format (not sure what OCA uses for this) with something like https://identity.foundation/wallet-rendering/

File: packages/openid4vc-holder/src/issuance/utils/IssuerMetadataUtils.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+): CredentialSupported[] {
+  const v8FormatEntries = Object.entries(supportedV8.formats)
+
+  return v8FormatEntries.map((entry) => {
+    const format = entry[0]
+    const credentialSupportBrief = entry[1]
+    if (typeof format !== 'string') {
+      throw Error(`Unknown format received ${JSON.stringify(format)}`)
+    }
+    let credentialSupport: Partial<CredentialSupported> = {}
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/issuance/utils/IssuerMetadataUtils.ts:164]: We assign a partial type here and later on cast to be fully typed. Can we get rid of the casting?

File: packages/openid4vc-holder/src/issuance/utils/IssuerMetadataUtils.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+export function handleLocations(authorizationDetails: AuthDetails, metadata: EndpointMetadataResult) {
+  if (metadata.credentialIssuerMetadata?.authorization_server || metadata.authorization_endpoint) {
+    if (!authorizationDetails.locations) authorizationDetails.locations = metadata.issuer
+    else if (Array.isArray(authorizationDetails.locations)) authorizationDetails.locations.push(metadata.issuer)
+    else authorizationDetails.locations = [authorizationDetails.locations as string, metadata.issuer]
+  }
+  return authorizationDetails
+}
+
+// TODO
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/issuance/utils/IssuerMetadataUtils.ts:205]: Is this still to do? If so, can you add some context as to why?

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+  Buffer,
+  DidsApi,
+  getJwkClassFromKeyType,
+  getKeyFromVerificationMethod,
+  injectable,
+  TypedArrayEncoder,
+  W3cJsonLdVerifiablePresentation,
+  asArray,
+} from '@aries-framework/core'
+import { CheckLinkedDomain, OP, ResponseMode, SupportedVersion, VerificationMode } from '@sphereon/did-auth-siop'
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:26]:
```suggestion
import { CheckLinkedDomain, OP as OpenIdProvider, ResponseMode, SupportedVersion, VerificationMode } from '@sphereon/did-auth-siop'
```

That is what it means, correct? OP seems quite vague.

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    request.presentationDefinitions.length === 1 &&
+    request.presentationDefinitions?.[0]?.definition !== undefined
+  )
+}
+
+@injectable()
+export class OpenId4VpHolderService {
+  public constructor(private presentationExchangeService: PresentationExchangeService) {}
+
+  private getOp(agentContext: AgentContext) {
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:51]: Same comment about `OP` being vague.

```suggestion
  private getOpenIdProvider(agentContext: AgentContext) {
```

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+        const inputDescriptor = credentialsForInputDescriptor[submission.inputDescriptorId]
+        if (!inputDescriptor) {
+          credentialsForInputDescriptor[submission.inputDescriptorId] = [verifiableCredential.credential]
+        } else {
+          inputDescriptor.push(verifiableCredential.credential)
+        }
+      })
+
+    const vps = await this.presentationExchangeService.createPresentation(agentContext, {
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:140]: Does `vps` mean verifiablePresentations?

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+    const response = await op.createAuthorizationResponse(options.verifiedAuthorizationRequest, {
+      issuer: verificationMethod.controller,
+      presentationExchange: {
+        verifiablePresentations: vps.verifiablePresentations.map((vp) => vp.encoded as W3CVerifiablePresentation),
+        presentationSubmission: vps.presentationSubmission,
+      },
+      signature: {
+        signature: async (data) => {
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:165]: Can't we type data here? Do we know the encoding of the data?

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+  private getSupportedDidMethods(agentContext: AgentContext) {
+    const didsApi = agentContext.dependencyManager.resolve(DidsApi)
+    const supportedDidMethods: string[] = []
+
+    for (const resolver of didsApi.config.resolvers) {
+      supportedDidMethods.push(...resolver.supportedMethods)
+    }
+
+    return supportedDidMethods
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:195]: Might be good to filter out duplicates here. Very minor comment though.

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  // TODO: we can do this in a simpler way, as we're now resolving it multiple times
+  private async getVerificationMethodFromVerifiablePresentation(
+    agentContext: AgentContext,
+    verifiablePresentation: W3cVerifiablePresentation
+  ) {
+    const didsApi = agentContext.dependencyManager.resolve(DidsApi)
+
+    let verificationMethod: string
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:219]:
```suggestion
    let verificationMethodId: string
```

File: packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+    }: {
+      credentialsForInputDescriptor: CredentialsForInputDescriptor
+      presentationDefinition: PresentationDefinitionV1 | PresentationDefinitionV2
+      challenge?: string
+      domain?: string
+      nonce?: string
+      includePresentationSubmissionInVp?: boolean
+    }
+  ) {
+    // if (selectedCredentials.length === 0) {
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts:100]: Remove commented code.

File: packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+      const credentialsForVp = Object.values(inputDescriptors)
+        .flatMap((inputDescriptors) => inputDescriptors)
+        .map(getSphereonW3cVerifiableCredential)
+
+      const presentationDefinitionForVp = {
+        ...presentationDefinition,
+        input_descriptors: inputDescriptorsForVp,
+
+        // We remove the submission requirements, as it will otherwise fail to create the VP
+        // FIXME: Will this cause issue for creating the credential? Need to run tests
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts:152]: Has this been tested?

File: packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+      descriptor_map: [],
+    }
+
+    for (const vp of verifiablePresentationResults) {
+      presentationSubmission.descriptor_map.push(
+        ...vp.presentationSubmission.descriptor_map.map((descriptor): Descriptor => {
+          const index = verifiablePresentationResults.indexOf(vp)
+          const prefix = verifiablePresentationResults.length > 1 ? `$[${index}]` : '$'
+          return {
+            format: 'jwt_vp',
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts:200]: Would be good to use an enum for this format and the one below in the `path_nested` object.

File: packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+      if (!alg) {
+        throw new AriesFrameworkError(`No supported algs for key type: ${key.keyType}`)
+      }
+
+      const signedPresentation = await w3cCredentialService.signPresentation(agentContext, {
+        format: ClaimFormat.JwtVp,
+        verificationMethod: verificationMethod.id,
+        presentation: w3cPresentation,
+        alg,
+        challenge: challenge ?? nonce ?? utils.uuid(),
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/PresentationExchangeService.ts:272]: we should not use a `uuid` as a challenge. `agenContext.wallet.generateNonce` can be used.

File: packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+      // FIXME: this is not collision proof as the name doesn't have to be unique
+      m.name === inputDescriptor.name
+  )
+
+  let name = inputDescriptor.name
+  // If there's no name on the input descriptor, but the id does not contain
+  // any special characters or numbers (so only letters and spaces),
+  // we will use a sanitized version of the id as the name
+  if (!name && inputDescriptor.id.match(/^[a-zA-Z ]+$/)) {
+    name = sanitizeString(inputDescriptor.id)
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts:287]: I think this is seomthing that should be done by the user and not the framework. Might miss some details as to why we do this though.

File: packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts [hunk 15]
```diff
@@ -0,0 +1,334 @@
+import type { PresentationSubmission, PresentationSubmissionRequirement, SubmissionEntry } from './types'
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts:1]: I don't think there is anythinng we can do about this, but this file contains quite some complex code (and quite some TODO's). Would be nice if we can clean this up later on.

File: packages/openid4vc-holder/src/presentations/transform.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+  ClaimFormat,
+  JsonTransformer,
+  AriesFrameworkError,
+  W3cJsonLdVerifiablePresentation,
+  W3cJwtVerifiablePresentation,
+  W3cJwtVerifiableCredential,
+  W3cJsonLdVerifiableCredential,
+} from '@aries-framework/core'
+
+export type { SphereonW3cVerifiableCredential, SphereonW3cVerifiablePresentation }
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/transform.ts:17]: Are these exposed to the AFJ users?

File: packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+import { AskarModule } from '@aries-framework/askar'
+import { KeyType, W3cJwtVerifiableCredential, Agent, Buffer } from '@aries-framework/core'
+import { agentDependencies } from '@aries-framework/node'
+import { ariesAskar } from '@hyperledger/aries-askar-nodejs'
+
+import { OpenId4VcHolderModule, OpenId4VpHolderService } from '../src'
+
+const modules = {
+  openId4VcClient: new OpenId4VcHolderModule(),
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts:11]:
```suggestion
  openId4VcHolder: new OpenId4VcHolderModule(),
```

File: packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+  askar: new AskarModule({
+    ariesAskar,
+  }),
+}
+
+describe('OpenId4VcClient | OpenID4VP', () => {
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts:17]:
```suggestion
describe('OpenId4VcHolder | OpenID4VP', () => {
```

File: packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+  })
+
+  afterEach(async () => {
+    await agent.shutdown()
+    await agent.wallet.delete()
+  })
+
+  describe('Mattr interop', () => {
+    // Not working yet. Once it works, we can mock the requests/responses
+    xit('Should succesfuly share a proof with MATTR launchpad', async () => {
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts:43]: Can we enable this?

File: packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+        ),
+      })
+
+      await agent.wallet.createKey({
+        keyType: KeyType.Ed25519,
+        seed: Buffer.from('00000000000000000000000000000000'),
+      })
+
+      await agent.dids.import({
+        did: 'did:key:z6MktqtXNG8CDUY9PrrtoStFzeCnhpMmgxYL1gikcW3BzvNW',
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts:57]: If this is the did associated with the generated key form above, you can do the following:

```typescript
const key = await agent.wallet.createKey({
  keyType: KeyType.Ed25519,
  seed: Buffer.from('00000000000000000000000000000000'),
})

const did = new DidKey(key)
```

File: packages/openid4vc-issuer/package.json [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+    "prepublishOnly": "yarn run build",
+    "test": "jest"
+  },
+  "dependencies": {
+    "@aries-framework/core": "0.4.2",
+    "@sphereon/oid4vci-issuer": "^0.7.3",
+    "@sphereon/oid4vci-common": "^0.7.3",
+    "@sphereon/ssi-types": "^0.17.5",
+    "@stablelib/random": "^1.0.2",
+    "fast-text-encoding": "^1.0.6"
```
Comment (berendsliedrecht) [inline packages/openid4vc-issuer/package.json:32]: Same comment regarding the fast-text-encoding and random.

File: packages/openid4vc-issuer/src/OpenId4VcIssuerApi.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
+import type { IssueCredentialOptions, SendCredentialOfferOptions } from './OpenId4VcIssuerServiceOptions'
```
Comment (berendsliedrecht) [inline packages/openid4vc-issuer/src/OpenId4VcIssuerApi.ts:1]: Might be good to include this in a separate PR. Would not want to confuse the users as the functionality is not here, yet.

File: packages/openid4vc-issuer/tests/openid4vc-issuer.e2e.test.ts [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+  DidKey,
+} from '@aries-framework/core'
+import { agentDependencies } from '@aries-framework/node'
+import { ariesAskar } from '@hyperledger/aries-askar-nodejs'
+import nock, { cleanAll, enableNetConnect } from 'nock'
+
+import { OpenId4VcIssuerModule } from '../src'
+
+const modules = {
+  openId4VcHolder: new OpenId4VcIssuerModule(),
```
Comment (berendsliedrecht) [inline packages/openid4vc-issuer/tests/openid4vc-issuer.e2e.test.ts:19]:
```suggestion
  openId4VcIssuer: new OpenId4VcIssuerModule(),
```

File: packages/openid4vc-verifier/src/OpenId4VcVerifierApi.ts [hunk 24]
```diff
@@ -0,0 +1,27 @@
+import type { IssueCredentialOptions, SendCredentialOfferOptions } from './OpenId4VcVerifierServiceOptions'
```
Comment (berendsliedrecht) [inline packages/openid4vc-verifier/src/OpenId4VcVerifierApi.ts:1]: Same goes for the verifier.

**Review by berendsliedrecht — CHANGES_REQUESTED** [review]
Great work @2mau!!! I think there is some cleanup to do, but thats always expected with such a big PR.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts [hunk 25]
```diff
@@ -0,0 +1,703 @@
...
+  // Scope and authorization_details can be used in the same authorization request
+  // https://datatracker.ietf.org/doc/html/draft-ietf-oauth-rar-23#name-relationship-to-scope-param
+  if (!nonEmptyScope && !nonEmptyAuthDetails) {
+    throw new AriesFrameworkError('Please provide a scope or authorization_details')
+  }
+
+  // Authorization servers supporting PAR SHOULD include the URL of their pushed authorization request endpoint in their authorization server metadata document
+  // Note that the presence of pushed_authorization_request_endpoint is sufficient for a client to determine that it may use the PAR flow.
+  let parEndpoint = metadata.credentialIssuerMetadata?.pushed_authorization_request_endpoint
+  if (typeof parEndpoint !== 'string') parEndpoint = undefined
```
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:108]: it is not required right?

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts [hunk 26]
```diff
@@ -0,0 +1,703 @@
...
+    resolvedCredentialOffer: ResolvedCredentialOffer,
+    authCodeFlowOptions: AuthCodeFlowOptions
+  ): Promise<ResolvedAuthorizationRequest> {
+    const { credentialOfferPayload, metadata: _metadata } = resolvedCredentialOffer
+    const { metadata } = await getMetadataFromCredentialOffer(credentialOfferPayload, _metadata)
+    const codeVerifier = randomStringForEntropy(256)
+    const codeVerifierSha256 = Hasher.hash(TypedArrayEncoder.fromString(codeVerifier), 'sha2-256')
+    const codeChallenge = TypedArrayEncoder.toBase64URL(codeVerifierSha256)
+
+    // TODO: authdetails
```
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:248]: One last thing I am unsure there is about https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-authorization-request
Basically users can request credentials with authorization_details or scopes
The mapping between scopes and credentials is out of scope.
In theory authorization details and scopes can be used in the same request,
and the authorization details param can be constructed automatically based on the requested credentials.
Since as far as I understood it all credentials can be requested via auth details it makes sense to only support auth details not scopes. But I am unsure if there are use cases for scopes. Should we just accept  both and send the provided scopes along the automatically generated authorization_details?
Or what is your opinion on this?

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts [hunk 27]
```diff
@@ -0,0 +1,703 @@
...
+      let credentialTypes: string | string[]
+      if (version < OpenId4VCIVersion.VER_1_0_11) {
+        if (isInlineOffer) throw new AriesFrameworkError(`Inline credential offers not supported for version < 11`)
+        if (!credentialWithMetadata.credentialSupported.id) {
+          throw new AriesFrameworkError( // This should not happen
+            `No id provided for a credential supported entry in combination with the OpenId4VCI v8 draft`
+          )
+        }
+        credentialTypes = credentialWithMetadata.credentialSupported.id.split(`-${originalFormat}`)[0]
+      } else {
```
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:408]: There is a check ensuring that the version is smaller then 11 right?
If the version is 11 this should not be required as the metadata is structured differently.

File: packages/openid4vc-holder/src/issuance/utils/Formats.ts [hunk 28]
```diff
@@ -0,0 +1,39 @@
+import type { OID4VCICredentialFormat } from '@sphereon/oid4vci-common'
+import type { CredentialFormat } from '@sphereon/ssi-types'
+
+import { OpenId4VCIVersion } from '@sphereon/oid4vci-common'
+
+// Base on https://github.com/Sphereon-Opensource/OID4VCI/pull/54/files
```
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/utils/Formats.ts:6]: nope but if they merge the PR at some point, we can remove this code as it would be contained in the lib.

File: packages/openid4vc-holder/src/index.ts [hunk 29]
```diff
@@ -0,0 +1,16 @@
+import 'fast-text-encoding'
```
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/index.ts:1]: Okay, so I discussed this with Timo. No one is really sure why this is here, so I am removing it for now.
Tested issuance on mobile and node, and it works without it.
But we may need to add this back in at a later point if required.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts [hunk 30]
```diff
@@ -0,0 +1,703 @@
...
+      credentialOfferPayload,
+      issuerMetadata,
+      version
+    )
+
+    const credentialsToRequestWithMetadata = credentialsToRequest?.map((ctr) => {
+      const credentialToRequest = offeredCredentialsWithMetadata.find((offeredCredentialWithMetadata) => {
+        const { format, types } = offeredCredentialWithMetadata
+        // only requests credentials with the exact same set of types and format
+        return ctr.format === format && ctr.types.sort().join(',') === types.sort().join(',')
```
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:347]: all I found is 'areObjectsEqual' and 'deepEquality'. Both do not work 👎  for this use case. Do you have any idea what the function is called? That helps me find it?

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (berendsliedrecht) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:248]: > Should we just accept both and send the provided scopes along the automatically generated authorization_details?

This seems like a good solution to me. `scopes` can still be used, but we let the user define this. 

Later on, we could create some kind of mapping a user can supply and go from there, but this should suffice for now.

It would be good to comment this code quite in detail as the behavior is a bit different.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (berendsliedrecht) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:347]: For only primitives you can export the `equalsIgnoreOrder` function, from the `deepEquality` file.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (berendsliedrecht) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:108]: I mean the typeof check.

the signature of `let parEndpoint = metadata.credentialIssuerMetadata?.pushed_authorization_request_endpoint` should be `string | undefined`, correct? If that is the case, the check below won't do anything unless someone intentionally tries to break it.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (berendsliedrecht) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:408]: Yes there is a check for smaller than 11. But if it is larger than 11 we will enter the `else` statement. If we have a newer version, let's say `13`, we will also enter the else statement, even if that only has worked for `11`.

[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:408]: thanks

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:108]: Ah, yes. I changed it. It was coded that way in the Sphereons libs. Not sure why. It seems they added the correct typings later. But it shouldn't break anything to change it.

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts (hunk 8 above)
Comment (2mau) [reply packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:195]: I am using a set instead of an array now

File: packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts (hunk 15 above)
Comment (2mau) [reply packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts:1]: Everything related to vp/siop is not yet finished. That explains the TODO's and some issues you mentioned in this review. Just the vci holder side is 'done'. So I will clean this up in the future.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:408]: Okay, now I understand. In my opinion, this is the best behavior for now. Since in the case of v > 11, let's say 13, I want to set the behavior of v11 as the default, as this is more likely to be the case right?

File: packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts (hunk 19 above)
Comment (2mau) [reply packages/openid4vc-holder/tests/openid4vp-holder.e2e.test.ts:43]: I will add it it the future

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts (hunk 6 above)
Comment (2mau) [reply packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:140]: i distructured the object for readability. Seemed also to me like verifiable presentations, but it contains multiple objects

File: packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts (hunk 14 above)
Comment (2mau) [reply packages/openid4vc-holder/src/presentations/selection/PexCredentialSelection.ts:287]: I ll leave a comment, and look at it later

File: packages/openid4vc-holder/src/presentations/transform.ts (hunk 16 above)
Comment (2mau) [reply packages/openid4vc-holder/src/presentations/transform.ts:17]: no

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts [hunk 31]
```diff
@@ -0,0 +1,703 @@
...
+
+  private signCallback(agentContext: AgentContext, verificationMethod: VerificationMethod) {
+    return async (jwt: Jwt, kid?: string) => {
+      if (!jwt.header) throw new AriesFrameworkError('No header present on JWT')
+      if (!jwt.payload) throw new AriesFrameworkError('No payload present on JWT')
+      if (!kid) throw new AriesFrameworkError('No KID is present in the callback')
+
+      // We have determined the verification method before and already passed that when creating the callback,
+      // however we just want to make sure that the kid matches the verification method id
+      if (verificationMethod.id !== kid) {
```
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:674]: We also set the kid to the verification method id. So this should always work.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (2mau) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:248]: This is one way to implement this.
The main problem at the moment are inline credential offers.
With inline credential offers, we don't have a context, we can't provide one.
However, we also may not know the scope, which I think is possible, since not every provider will create a custom mapping of scopes to credentials.
That would mean we can't request that credential at all :( 

So in my opinion just send the context as undefined in that cases for now?
Or what are your thoughts on this?

```ts
  private getAuthDetailsFromOfferedCredential(
    credentialWithMetadata: OfferedCredentialWithMetadata,
    authDetailsLocation: string | undefined,
    version: OpenId4VCIVersion
  ): AuthDetails {
    const { format, types } = credentialWithMetadata
    const type = 'openid_credential'

    if (version < OpenId4VCIVersion.VER_1_0_11) {
      const credential_type = getV8CredentialType(credentialWithMetadata, format, version)
      return { type, credential_type, format }
    }

    const locations = authDetailsLocation ? [authDetailsLocation] : undefined
    if (format === OpenIdCredentialFormatProfile.JwtVcJson || format === OpenIdCredentialFormatProfile.JwtVcJsonLd) {
      return {
        type,
        format,
        types,
        locations,
      }
    } else if (format === OpenIdCredentialFormatProfile.LdpVc) {
      let context: string | undefined = undefined

      if (credentialWithMetadata.offerType === OfferedCredentialType.InlineCredentialOffer) {
        // Inline Credential Offers come with no context
      } else {
        if ('@context' in credentialWithMetadata.credentialSupported) {
          context = credentialWithMetadata.credentialSupported['@context'] as unknown as string
        } else {
          throw new AriesFrameworkError('Could not find @context in credentialSupported.')
        }
      }

      return {
        type,
        format,
        types,
        locations,
        '@context': context,
      }
    } else {
      throw new AriesFrameworkError(`Cannot create authorization_details. Unsupported credential format ${format}.`)
    }
  }
```

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (berendsliedrecht) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:493]: IMO it would be better to error out as a lot of undefined behavior can happen because of this. It is exactly "likely" that it is correct, but not 100%.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (berendsliedrecht) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts]: What are the implications of returning an undefined context? Would not just make more sense for that unique scenario to just error out for now? If there is a need from the users we can check it again, but it seems quite niche that it might be a waste of time to get it perfect.

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (berendsliedrecht) [reply packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:493]: I think almost all code has a fallback to v11 so it does not really matter if we change this here, you can leave it as-is.

File: packages/openid4vc-holder/src/OpenId4VcHolderApi.ts [hunk 32]
```diff
@@ -0,0 +1,155 @@
...
+    redirect_url: string
+    clientMetadata?: ClientMetadataOpts
+    issuer?: string
+  }) {
+    const relyingParty = await this.openId4VpHolderService.getRelyingParty(this.agentContext, options)
+
+    // TODO: generate nonce, state, correlationId
+    const nonce = 'qBrR7mqnY3Qr49dAZycPF8FzgE83m6H0c2l0bzP4xSg'
+    const state = 'b32f0087fc9816eb813fd11f'
+    const correlationId = '1'
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/OpenId4VcHolderApi.ts:46]: This should be fixed before merging, correct?

File: packages/openid4vc-holder/src/OpenId4VcHolderApi.ts [hunk 33]
```diff
@@ -0,0 +1,155 @@
...
+  ): Promise<W3cCredentialRecord[]> {
+    return this.openId4VcHolderService.acceptCredentialOffer(this.agentContext, {
+      resolvedCredentialOffer,
+      acceptCredentialOfferOptions,
+    })
+  }
+
+  /**
+   * Accepts a credential offer using the authorization code flow.
+   * @param resolvedCredentialOffer Obtained through @function resolveCredentialOffer
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/OpenId4VcHolderApi.ts:137]: `@function` does not seem like valid ts-doc. Maybe check the `@link`?

File: packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts [hunk 34]
```diff
@@ -0,0 +1,782 @@
...
+      authorizationRequestUri,
+    }
+  }
+
+  public async acceptCredentialOffer(
+    agentContext: AgentContext,
+    options: {
+      resolvedCredentialOffer: ResolvedCredentialOffer
+      acceptCredentialOfferOptions: AcceptCredentialOfferOptions
+      resolvedAuthorizationRequest?: ResolvedAuthorizationRequest & { code: string }
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/issuance/OpenId4VciHolderService.ts:361]: Can't we add the `{code:string}` to the `ResolvedAuthorizationRequest`? Or create a new type?

File: packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts [hunk 35]
```diff
@@ -0,0 +1,524 @@
+import type { PresentationSubmission } from './selection'
```
Comment (berendsliedrecht) [inline packages/openid4vc-holder/src/presentations/OpenId4VpHolderService.ts:1]: Ah I wasn't aware the this implementation of vp. I think if there is some code already there it would be good to add this in a separate PR.

File: packages/openid4vc-issuer/src/index.ts [hunk 36]
```diff
@@ -0,0 +1,6 @@
+export * from './OpenId4VcIssuerApi'
+export * from './OpenId4VcIssuerModule'
+export * from './OpenId4VcIssuerService'
+
+// Contains internal types, so we don't export everything
+export {} from './OpenId4VcIssuerServiceOptions'
```
Comment (berendsliedrecht) [inline packages/openid4vc-issuer/src/index.ts:6]: We can just omit this line, right?

File: packages/openid4vc-holder/src/OpenId4VcHolderApi.ts (hunk 32 above)
Comment (2mau) [reply packages/openid4vc-holder/src/OpenId4VcHolderApi.ts:46]: Yes
